File: databasyrepo/databasyrepo/models/core/events.py
from databasyrepo.models.core.serializing import Serializable
from databasyrepo.models.core.actions import Register, Unregister, Set, InsertItem, DeleteItem

__author__ = 'Marboni'

class Event(Serializable):
    def do_action(self):
        """ Returns action to make this modification.
        """
        raise NotImplementedError

    def undo_action(self):
        """ Returns action to revert this modification.
        """
        raise NotImplementedError

class NodeRegistered(Event):
    def __init__(self, node=None):
        super(NodeRegistered, self).__init__()
        if node:
            self['node'] = node.copy()

    def fields(self):
        from databasyrepo.models.core.nodes import Node
        return {
            'node': Node
        }

    def do_action(self):
        return Register(node=self['node'])

    def undo_action(self):
        return Unregister(node_id=self['node'].id)


class NodeUnregistered(Event):
    def __init__(self, node=None):
        super(NodeUnregistered, self).__init__()
        if node:
            self['node'] = node.copy()

    def fields(self):
        from databasyrepo.models.core.nodes import Node
        return {
            'node': Node
        }

    def do_action(self):
        return Unregister(node_id=self['node'].id)

    def undo_action(self):
        return Register(node=self['node'])


class PropertyChanged(Event):
    def fields(self):
        return {
            'node_id': basestring,
            'field': basestring,
            'old_value': object,
            'new_value': object
        }

    def do_action(self):
        return Set(node_id=self['node_id'], field=self['field'], value=self['new_value'])

    def undo_action(self):
        return Set(node_id=self['node_id'], field=self['field'], value=self['old_value'])

class ItemInserted(Event):
    def fields(self):
        return {
            'node_id': basestring,
            'field': basestring,
            'index': object,
            'item': object
        }

    def do_action(self):
        return InsertItem(node_id=self['node_id'], field=self['field'], index=self['index'], item=self['item'])

    def undo_action(self):
        return DeleteItem(node_id=self['node_id'], field=self['field'], index=self['index'])


class ItemDeleted(Event):
    def fields(self):
        return {
            'node_id': basestring,
            'field': basestring,
            'index': object,
            'item': object
        }

    def do_action(self):
        return DeleteItem(node_id=self['node_id'], field=self['field'], index=self['index'])

    def undo_action(self):
        return InsertItem(node_id=self['node_id'], field=self['field'], index=self['index'], item=self['item'])

File: databasyrepo/databasyrepo/static/repo/src/databasy/model/core/events.js
databasy.model.core.events.Event = databasy.model.core.serializing.Serializable.extend({
    do_action:function() {
        throw new Error('Not implemented.');
    },
    undo_action:function() {
        throw new Error('Not implemented.');
    }
});

databasy.model.core.events.NodeRegistered = databasy.model.core.events.Event.extend({
    init:function(node) {
        this._super();
        if (node !== undefined) {
            this.set('node', node.copy())
        }
    },
    fields:function() {
        return this._super().concat(
            'node'
        )
    },
    do_action:function() {
        return databasy.model.core.actions.Register({node: this.val('node')});
    },
    undo_action:function() {
        return databasy.model.core.actions.Unregister({node_id: this.val('node').id()});
    }
}, {
    CODE: 'core.events.NodeRegistered'
});


databasy.model.core.events.NodeUnregistered = databasy.model.core.events.Event.extend({
    init:function(node) {
        this._super();
        if (node !== undefined) {
            this.set('node', node.copy())
        }
    },
    fields:function() {
        return this._super().concat(
            'node'
        )
    },
    do_action:function() {
        return databasy.model.core.actions.Unregister({node_id: this.val('node').id()});
    },
    undo_action:function() {
        return databasy.model.core.actions.Register({node: this.val('node')});
    }
}, {
    CODE: 'core.events.NodeUnregistered'
});


databasy.model.core.events.PropertyChanged = databasy.model.core.events.Event.extend({
    fields:function() {
        return this._super().concat(
            'node_id',
            'field',
            'old_value',
            'new_value'
        )
    },
    do_action:function() {
        return databasy.model.core.actions.Set({node_id: this.val('node_id'), field:this.val('field'), value:this.val('new_value')});
    },
    undo_action:function() {
        return databasy.model.core.actions.Set({node_id: this.val('node_id'), field:this.val('field'), value:this.val('old_value')});
    }
}, {
    CODE: 'core.events.PropertyChanged'
});


databasy.model.core.events.ItemInserted = databasy.model.core.events.Event.extend({
    fields:function() {
        return this._super().concat(
            'node_id',
            'field',
            'index',
            'item'
        )
    },
    do_action:function() {
        return databasy.model.core.actions.InsertItem({node_id: this.val('node_id'), field:this.val('field'), index:this.val('index'), item:this.val('item')});
    },
    undo_action:function() {
        return databasy.model.core.actions.DeleteItem({node_id: this.val('node_id'), field:this.val('field'), index:this.val('index')});
    }
}, {
    CODE: 'core.events.ItemInserted'
});


databasy.model.core.events.ItemDeleted = databasy.model.core.events.Event.extend({
    fields:function() {
        return this._super().concat(
            'node_id',
            'field',
            'index',
            'item'
        )
    },
    do_action:function() {
        return databasy.model.core.actions.DeleteItem({node_id: this.val('node_id'), field:this.val('field'), index:this.val('index')});
    },
    undo_action:function() {
        return databasy.model.core.actions.InsertItem({node_id: this.val('node_id'), field:this.val('field'), index:this.val('index'), item:this.val('item')});
    }
}, {
    CODE: 'core.events.ItemDeleted'
});

